clippy_lints/src/methods.rs

```diff
@@ -1398,6 +1403,45 @@ impl SelfKind {
     }
 }
 
+fn is_astrait(ty: &hir::Ty, self_ty: &hir::Ty, generics: &hir::Generics, name: &[&str]) -> bool {
+    single_segment_ty(ty).map_or(false, |seg| {
+        generics.ty_params.iter().any(|param| {
+            param.name == seg.name && param.bounds.iter().any(|bound| {
+                if let hir::TyParamBound::TraitTyParamBound(ref ptr, ..) = *bound {
+                    let path = &ptr.trait_ref.path;
+                    match_path_old(path, name) && path.segments.last().map_or(false, |s| {
+                        if let hir::PathParameters::AngleBracketedParameters(ref data) = s.parameters {
+                            data.types.len() == 1 && (is_self_ty(&data.types[0]) || is_ty(&*data.types[0], self_ty))
+                        } else {
+                            false
+                        }
+                    })
+                } else {
+                    false
+                }
+            })
+        })
+    })
+}
+
+fn is_ty(ty: &hir::Ty, self_ty: &hir::Ty) -> bool {
+    match (&ty.node, &self_ty.node) {
+        (&hir::TyPath(hir::QPath::Resolved(_, ref ty_path)), &hir::TyPath(hir::QPath::Resolved(_, ref self_ty_path))) => {
+            ty_path.segments.iter().rev().map(|seg| seg.name).zip(
+                self_ty_path.segments.iter().rev().map(|seg| seg.name)).all(|(l, r)| l == r)
```

Comment:
This should use Iterator's eq method. And we need an issue open about linting this pattern ;)